### DISCUSSION TITLE
Does pmd still support command line to scan java files?

### DISCUSSION THREAD

**Discussion by vivian196167** [pmd/pmd#4029] [General]
Hi,

I checked from PMD official website https://pmd.github.io/pmd-6.47.0/index.html#-documentation, it looks like pmd commd line can check java files, this comes from the website:
~ $ cd ~/bin/pmd-bin-6.47.0/bin
~/.../bin $ ./run.sh pmd -d ../../../src/main/java/ -f text -R rulesets/java/quickstart.xml
  
  .../src/main/java/com/me/RuleSet.java:123  These nested if statements could be combined
  .../src/main/java/com/me/RuleSet.java:231  Useless parentheses.
  .../src/main/java/com/me/RuleSet.java:232  Useless parentheses.
  .../src/main/java/com/me/RuleSet.java:357  These nested if statements could be combined
  .../src/main/java/com/me/RuleSetWriter.java:66     Avoid empty catch blocks

But I checked in my ubundu16.04, it seems not support java any more? can someone help me figure it out?

run.sh pmd -d app -f text -R etc/pmd/relest.xml 
Ignoring language for PMD: java.util.ServiceConfigurationError: net.sourceforge.pmd.lang.Language: Provider net.sourceforge.pmd.lang.java.JavaLanguageModule could not be instantiated
Ignoring language for PMD: java.util.ServiceConfigurationError: net.sourceforge.pmd.lang.Language: Provider net.sourceforge.pmd.lang.java.JavaLanguageModule could not be instantiated
Jun 30, 2022 4:50:55 PM net.sourceforge.pmd.util.log.internal.SimpleMessageReporter logImpl
SEVERE: Cannot load ruleset etc/pmd/relest.xml: Unknown Language 'java' for rule ThreadPoolCreationRule, supported Languages are apex, ecmascript, html, jsp, modelica, plsql, pom, scala, vf, vm, wsdl, xml, xsl
Jun 30, 2022 4:50:55 PM net.sourceforge.pmd.util.log.internal.SimpleMessageReporter logImpl
WARNING: No rules found. Maybe you misspelled a rule name? (etc/pmd/relest.xml)

Thanks very much

**Comment by adangel** [comment]
Thanks for asking. Java should of course be supported by PMD.

Can you please run with the `--debug` flag, e.g. `run.sh pmd -d app -f text -R etc/pmd/relest.xml` ?
Maybe we see, why "JavaLanguageModule could not be instantiated"...

Also, which version of java are you using? Please also post the output of `java -version` and `echo $JAVA_HOME` and `$JAVA_HOME/bin/java -version`.

**Comment by vivian196167** [comment]
Thanks Adangel for quick response.

java -version
openjdk version "1.8.0_282"
OpenJDK Runtime Environment (build 1.8.0_282-8u282-b08-0ubuntu1~16.04-b08)
OpenJDK 64-Bit Server VM (build 25.282-b08, mixed mode)

I did not set $JAVA_HOME， I have multiple java in my env and I choose to use java 1.8.
sudo update-alternatives --config java
There are 2 choices for the alternative java (providing /usr/bin/java).
， I have multiple java in my env and I choose to use java 1.8.

```
sudo update-alternatives --config java
There are 2 choices for the alternative java (p
  Selection    Path                                            Priority   Status
------------------------------------------------------------
  0            /usr/lib/jvm/java-8-openjdk-amd64/jre/bin/java   1081      auto mode
  1            /usr/jdk-11.0.15.1/bin/java                      2         manual mode
* 2            /usr/lib/jvm/java-8-openjdk-amd64/jre/bin/java   1081      manual mode
```

I am using pmd-bin-6.47.0, and copied p3c-pmd-2.1.1.jar  under its libs, which can  compile from  https://github.com/alibaba/p3c/tree/master/p3c-pmd. sorry I used company's laptop  which cannot upload any file to network, will violate company's security strategy.

relest.xml is like this:

```xml
<?xml version="1.0"?>
<ruleset name="Custom ruleset"
         xmlns="http://pmd.sourceforge.net/ruleset/2.0.0"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://pmd.sourceforge.net/ruleset/2.0.0 http://pmd.sourceforge.net/ruleset_2_0_0.xsd">
    <description>
        自定义Rule set
    </description>
    <!-- PMD Rules, From https://github.com/pmd/pmd/tree/master/pmd-java/src/main/resources/rulesets/java  -->
    <rule ref="rulesets/java/android.xml">
        <exclude name="CallSuperLast"/>
    </rule>
    <rule ref="rulesets/java/basic.xml">
        <exclude name="CollapsibleIfStatements"/>
    </rule>
    <rule ref="rulesets/java/clone.xml"/>
    <rule ref="rulesets/java/finalizers.xml"/>
    <rule ref="rulesets/java/imports.xml"/>
    <rule ref="rulesets/java/javabeans.xml"/>

    <rule ref="rulesets/java/optimizations.xml">
        <exclude name="LocalVariableCouldBeFinal"/>
        <exclude name="MethodArgumentCouldBeFinal"/>
    </rule>
    <rule ref="rulesets/java/sunsecure.xml"/>

    <rule ref="rulesets/java/unnecessary.xml">
        <exclude name="UselessParentheses"/>
    </rule>

    <!-- Ali Rules, From https://github.com/alibaba/p3c/tree/master/p3c-pmd/src/main/resources/rulesets/java -->
    <rule ref="rulesets/java/ali-comment.xml">
    </rule>

    <rule ref="rulesets/java/ali-concurrent.xml">
    </rule>

    <rule ref="rulesets/java/ali-constant.xml">
    </rule>

    <rule ref="rulesets/java/ali-exception.xml">
    </rule>

    <rule ref="rulesets/java/ali-flowcontrol.xml">
    </rule>

    <rule ref="rulesets/java/ali-naming.xml">
    </rule>

    <rule ref="rulesets/java/ali-oop.xml">
    </rule>

    <rule ref="rulesets/java/ali-orm.xml">
    </rule>

    <rule ref="rulesets/java/ali-other.xml">
    </rule>

    <rule ref="rulesets/java/ali-set.xml">
    </rule>

</ruleset>
```

**Reply by adangel** [reply]
Please run the following commands and post the complete output of the commands here:

1. `run.sh run.sh pmd --help`
2. `ls -l "$(dirname $(which run.sh))/../lib"`

I can't reproduce your error - and the problem about the ruleset is just a consequential error, since the PMD's java language couldn't be loaded and all enabled rules would require this language - no rules are left.
I suspect some classloader issues. I can reproduce a similar error, if I use [p3c-pmd-2.1.1-jar-with-dependencies.jar](https://repo1.maven.org/maven2/com/alibaba/p3c/p3c-pmd/2.1.1/p3c-pmd-2.1.1-jar-with-dependencies.jar) instead of [p3c-pmd-2.1.1.jar](https://repo1.maven.org/maven2/com/alibaba/p3c/p3c-pmd/2.1.1/p3c-pmd-2.1.1.jar). But the error should be printed already with the first command. For me it looks like this:

```
$ run.sh pmd --help
Usage: pmd [options]
  Options:
    -failOnViolation, --failOnViolation
      By default PMD exits with status 4 if violations are found. Disable this
...
Exception in thread "main" java.util.ServiceConfigurationError: net.sourceforge.pmd.lang.Language: Provider net.sourceforge.pmd.lang.cpp.CppLanguageModule could not be instantiated
	at java.util.ServiceLoader.fail(ServiceLoader.java:232)
	at java.util.ServiceLoader.access$100(ServiceLoader.java:185)
	at java.util.ServiceLoader$LazyIterator.nextService(ServiceLoader.java:384)
	at java.util.ServiceLoader$LazyIterator.next(ServiceLoader.java:404)
	at java.util.ServiceLoader$1.next(ServiceLoader.java:480)
	at net.sourceforge.pmd.lang.LanguageRegistry.<init>(LanguageRegistry.java:35)
	at net.sourceforge.pmd.lang.LanguageRegistry.<clinit>(LanguageRegistry.java:24)
	at net.sourceforge.pmd.cli.PMDCommandLineInterface.supportedVersions(PMDCommandLineInterface.java:122)
	at net.sourceforge.pmd.cli.PMDCommandLineInterface.buildUsageText(PMDCommandLineInterface.java:77)
	at net.sourceforge.pmd.cli.PMDCommandLineInterface.extractParameters(PMDCommandLineInterface.java:48)
	at net.sourceforge.pmd.PMD.run(PMD.java:468)
	at net.sourceforge.pmd.cli.PMDCommandLineInterface.run(PMDCommandLineInterface.java:165)
	at net.sourceforge.pmd.PMD.main(PMD.java:455)
Caused by: java.lang.NoSuchMethodError: net.sourceforge.pmd.lang.BaseLanguageModule.<init>(Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;[Ljava/lang/String;)V
	at net.sourceforge.pmd.lang.cpp.CppLanguageModule.<init>(CppLanguageModule.java:27)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at java.lang.Class.newInstance(Class.java:442)
	at java.util.ServiceLoader$LazyIterator.nextService(ServiceLoader.java:380)
	... 10 more
```

**Comment by vivian196167** [comment]
Can command line support multile rulesets or how to group multiple rulesets to make it happen?

**Reply by oowekyala** [reply]
You can use the `-R` option several times, or provide several arguments to a single `-R` option (since pmd 6.46.0)
